File: 0337-house-robber-iii/0337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class RobbedRoot {
public:
    int robbedRoot;
    int skippedRoot;
    RobbedRoot(int a,int b){
        robbedRoot = a;
        skippedRoot = b;
    }
};

class Solution {
private:
    RobbedRoot dfs(TreeNode* root){
        if(root == nullptr){
            return RobbedRoot(0,0);
        }

        RobbedRoot robLeft = dfs(root->left);
        RobbedRoot robRight = dfs(root->right);

        int robThis = root->val + robLeft.skippedRoot + robRight.skippedRoot;
        int skipThis = robLeft.robbedRoot + robRight.robbedRoot;

        return RobbedRoot(robThis,skipThis);

    }
public:
    int rob(TreeNode* root) {
        RobbedRoot res = dfs(root);
        return max(res.robbedRoot , res.skippedRoot);
    }
};

File: 0337-house-robber-iii/0337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root: TreeNode) -> List[int]:
        if not root:
            return [0,0]
        
        left = self.solve(root.left)
        right = self.solve(root.right)

        res = [0,0]

        res[0] = max(left[0], left[1]) + max(right[0],right[1])
        res[1] = root.val + left[0] + right[0]

        return res


    def rob(self, root: Optional[TreeNode]) -> int:
        arr = self.solve(root)
        return max(arr[0],arr[1])